daedaland/AthanasiaOtc.sol

```diff
@@ -0,0 +1,811 @@
+     *  - `expectedCost` - total number of `otcToken` tokens that the caller expects to pay for the `amountToPurchase` underlying tokens.
+     */
+    function otc(
+        address collection,
+        uint256 amountToPurchase,
+        uint256 expectedCost
+    ) external payable;
+}
+
+interface IStakingHelper {
+    function stake(uint256 _amount, address _recipient) external;
+}
+
+interface IHectorTreasury {
+    function mintRewards(address _recipient, uint256 _amount) external;
+}
+
+contract AthanasiaOtc is IAthanasiaOtc, Ownable, Pausable {
+    using SafeERC20 for IERC20;
+
+    uint256 constant PRICE_MULTIPLIER = 10**9;
+
+    IERC20 public immutable HEC;
+    IERC20 public immutable sHEC;
+    IStakingHelper public immutable stakingHelper;
+    IHectorTreasury public immutable hectorTreasury;
+
+    mapping(address => TokenInfo) collections;
+    mapping(address => mapping(address => uint256)) userPurchasedForCollection;
+
+    constructor(
+        address _HEC,
+        address _sHEC,
+        address _stakingHelper,
+        address _hectorTreasury
+    ) {
+        require(_HEC != address(0), 'OTC: Invalid HEC');
+        require(_sHEC != address(0), 'OTC: Invalid sHEC');
+        require(_stakingHelper != address(0), 'OTC: Invalid staking helper');
+        require(_hectorTreasury != address(0), 'OTC: Invalid hector treasury');
+
+        HEC = IERC20(_HEC);
+        sHEC = IERC20(_sHEC);
+        stakingHelper = IStakingHelper(_stakingHelper);
+        hectorTreasury = IHectorTreasury(_hectorTreasury);
+    }
+
+    modifier onlyExistingCollection(address collection) {
+        require(
+            collections[collection].totalAmount > 0,
+            'OTC: Unregistered collection'
+        );
+        _;
+    }
+
+    function registerCollection(
+        address collection,
+        address otcToken,
+        uint256 otcPrice,
+        uint256 totalAmount
+    ) external override onlyOwner {
+        require(collection != address(0), 'OTC: Invalid collection');
+        require(otcPrice > 0, 'OTC: Invalid otc price');
+        require(totalAmount > 0, 'OTC: Invalid total amount');
+
+        TokenInfo storage info = collections[collection];
+        info.otcToken = otcToken;
+        info.otcPrice = otcPrice;
+        info.totalAmount = totalAmount;
+    }
+
+    function validateCollection(
+        address collection,
+        address otcToken,
+        uint256 otcPrice
+    ) external view override returns (bool) {
+        TokenInfo memory info = collections[collection];
+
+        return info.otcToken == otcToken && info.otcPrice == otcPrice;
+    }
+
+    function otc(
+        address collection,
+        uint256 amountToPurchase,
+        uint256 expectedCost
+    ) external payable override onlyExistingCollection(collection) {
+        TokenInfo memory info = collections[collection];
+        uint256 cost = (amountToPurchase * info.otcPrice) / PRICE_MULTIPLIER;
+
```

Comment:
right now, anyone can use the OTC purchase function for any existing collection, which is probably not what you want. only the IAthanasia contract should be allowed to do that (or if you also want to add this option, the NFT project itself, but I think we talked about this and decided only IAthanasia contract should be allowed to do the otc call on behalf of the NFT collection.
For this you probably want to add the IAthanasia otc contract address to the constructor and validate it in otc function

Reply:
> 

yeah, but there is no need to complicate it as it is in the TOR function you referenced, because we know exactly how many decimals HEC has, and the decimals from otcToken are not important, as otcPrice already capture the price of exactly 1 HEC.

Reply:
> > 
> 
> yeah, but there is no need to complicate it as it is in the TOR function you referenced, because we know exactly how many decimals HEC has, and the decimals from otcToken are not important, as otcPrice already capture the price of exactly 1 HEC.

So there's no need to fix the current price calculation